allauth/socialaccount/providers/apple/views.py

```diff
@@ -0,0 +1,130 @@
+import json
+import requests
+from datetime import timedelta
+
+from django.http import HttpResponseRedirect
+from django.utils import timezone
+from django.utils.http import urlencode
+from django.views.decorators.csrf import csrf_exempt
+
+import jwt
+
+from allauth.socialaccount.models import SocialApp, SocialToken
+from allauth.socialaccount.providers.oauth2.client import OAuth2Error
+from allauth.socialaccount.providers.oauth2.views import (
+    OAuth2Adapter,
+    OAuth2CallbackView,
+    OAuth2LoginView,
+)
+
+from .client import AppleOAuth2Client
+from .provider import AppleProvider
+
+
+class AppleOAuth2Adapter(OAuth2Adapter):
+    provider_id = AppleProvider.id
+    access_token_url = "https://appleid.apple.com/auth/token"
+    authorize_url = "https://appleid.apple.com/auth/authorize"
+    public_key_url = "https://appleid.apple.com/auth/keys"
+
+    def get_public_key(self, id_token):
+        """
+        Get the public key which matches the `kid` in the id_token header.
+        """
+        kid = jwt.get_unverified_header(id_token)["kid"]
+        apple_public_key = [
+            d
+            for d in requests.get(self.public_key_url).json()["keys"]
+            if d["kid"] == kid
+        ][0]
+        public_key = jwt.algorithms.RSAAlgorithm.from_jwk(
+            json.dumps(apple_public_key)
+        )
+        return public_key
+
+    def get_client_id(self, provider):
+        app = SocialApp.objects.get(provider=provider.id)
+        return [aud.strip() for aud in app.client_id.split(",")]
+
+    def parse_token(self, data):
+        try:
+            token = SocialToken(token=data["access_token"])
+            token.token_secret = data.get("refresh_token", "")
+
+            public_key = self.get_public_key(data["id_token"])
+            provider = self.get_provider()
+            allowed_auds = self.get_client_id(provider)
+
+            token.user_data = jwt.decode(
```

Comment:
The implementation of sign in with apple is weired. We are getting the JWT key form apple server, why do we need to verify again with the key they have provided? It makes the thing much slower unfortunately!

Reply:
Sure we have, it makes verification that no one has interfered to logging in.
That's how I understand the process. [here are the docs](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_rest_api/verifying_a_user#overview)
@safwanrahman

Reply:
@0leg5ergeev if anyone interfered, we would not get the access token while generating the token. That is what I understand. Correct me if I am wrong.

Reply:
@safwanrahman I think you're right, but as @0leg5ergeev points out the docs say you _must_ verify the signature. It does seem a little heavy-handed to verify it when it's come straight from an apple endpoint. 

I'd vote to keep it, because a) it follows the apple guidelines, b) it is easier for hooking in to the auth process via an API (In which case the id_token may be from an iOS SDK)

Reply:
@benhowes It was really confusing actually as per the docs. But the identity token that we should decode is passed in `id_token` key while the user [got redirected to our call back url](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_js/incorporating_sign_in_with_apple_into_other_platforms#3332115). As it is coming directly from the client, we need to validate the `id_token` when it is coming from the callback.